gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagProcessingEngine.java

```diff
@@ -0,0 +1,128 @@
+  private static final Integer DEFAULT_NUM_THREADS = 3;
+  private static final Integer DEFAULT_JOB_STATUS_POLLING_INTERVAL = 10;
+
+  private final DagTaskStream dagTaskStream;
+  Optional<EventSubmitter> eventSubmitter;
+
+  public DagProcessingEngine(Config config, DagTaskStream dagTaskStream, DagProcFactory dagProcFactory, MultiActiveLeaseArbiter multiActiveLeaseArbiter) {
+    MetricContext metricContext = Instrumented.getMetricContext(ConfigUtils.configToState(ConfigFactory.empty()), getClass());
+    this.eventSubmitter = Optional.of(new EventSubmitter.Builder(metricContext, "org.apache.gobblin.service").build());
+    Integer numThreads = ConfigUtils.getInt(config, NUM_THREADS_KEY, DEFAULT_NUM_THREADS);
+    ScheduledExecutorService scheduledExecutorPool = Executors.newScheduledThreadPool(numThreads);
+    Integer pollingInterval = ConfigUtils.getInt(config, JOB_STATUS_POLLING_INTERVAL_KEY, DEFAULT_JOB_STATUS_POLLING_INTERVAL);
+    this.dagTaskStream = dagTaskStream;
+
+    for (int i=0; i < numThreads; i++) {
+      DagProcEngineThread dagProcEngineThread = new DagProcEngineThread(dagTaskStream, dagProcFactory);
+      scheduledExecutorPool.scheduleAtFixedRate(dagProcEngineThread, 0, pollingInterval, TimeUnit.SECONDS);
+    }
+  }
+
+  public void addNewDag(Dag<JobExecutionPlan> dag) {
+    this.dagTaskStream.addDagAction(DagManagerUtils.createDagAction(dag, DagActionStore.FlowActionType.LAUNCH));
+  }
+
+  public void addDagNodeToRetry(Dag.DagNode<JobExecutionPlan> dagNode) {
+    // todo - how to add dag action for for a dag node? should we create a dag node action? right now dag action is synonymous to flow action
+    // this.dagTaskStream.addDagTask(new RetryDagTask(dagNode));
+  }
+
+  public void addKillDagAction(Dag<JobExecutionPlan> dag) {
+    this.dagTaskStream.addDagAction(DagManagerUtils.createDagAction(dag, DagActionStore.FlowActionType.KILL));
+  }
+
+  public void addKillDagAction(Dag.DagNode<JobExecutionPlan> dagNode) {
+    this.dagTaskStream.addDagAction(DagManagerUtils.createDagAction(dagNode, DagActionStore.FlowActionType.KILL));
+  }
+
+  public void addAdvanceDagAction(Dag.DagNode<JobExecutionPlan> dagNode) {
+    this.dagTaskStream.addDagAction(DagManagerUtils.createDagAction(dagNode, DagActionStore.FlowActionType.ADVANCE));
+  }
+
+  public void addDagAction(DagActionStore.DagAction dagAction) {
+    this.dagTaskStream.addDagAction(dagAction);
+  }
+
+  @AllArgsConstructor
+  private static class DagProcEngineThread implements Runnable {
+
+    private DagTaskStream dagTaskStream;
+    private DagProcFactory dagProcFactory;
+
+    @Override
+    public void run() {
+      for (DagTaskStream it = dagTaskStream; it.hasNext(); ) {
+        DagTask dagTask = it.next();
+        DagProc dagProc = dagTask.host(dagProcFactory);
+//          dagProc.process(eventSubmitter.get(), maxRetryAttempts, delayRetryMillis);
+        try {
+          // todo - add retries
+          dagProc.process(dagTask);
```

Comment:
since the `DagProc` was created from the `DagTask`, there should be no reason to pass the dag task once again to the dag proc created from it